internal/mapdecode/hooks.go

```diff
@@ -0,0 +1,177 @@
+			// let's process it gracefully anyway.
+			value = reflect.Zero(from)
+		}
+
+		out, err := hook(from, to, value)
+		if err != nil {
+			return nil, err
+		}
+
+		return out.Interface(), nil
+	}
+}
+
+// Compposes multiple reflectHooks into one. The hooks are applied in-order
+// and values produced by a hook are fed into the next hook.
+func composeReflectHooks(hooks ...reflectHook) reflectHook {
+	return func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		var err error
+		for _, hook := range hooks {
+			data, err = hook(from, to, data)
+			if err != nil {
+				return data, err
+			}
+
+			// Update the `from` type to reflect changes made by the hook.
+			from = data.Type()
+		}
+		return data, err
+	}
+}
+
+// Wraps a reflectHook to support pointers in either direction (source and
+// destination).
+func supportPointers(hook reflectHook) (outputHook reflectHook) {
+	outputHook = func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		// Get rid of pointers in either direction. This lets us parse **foo if we
+		// know how to parse foo.
+		switch {
+		case from == to:
+			return data, nil
+		case from.Kind() == reflect.Ptr: // *foo => bar
+			// Decoding a pointer type to a non-pointer type. Dereference if
+			// non-nil, use zero-value otherwise.
+			from = from.Elem()
+			if data.IsNil() {
+				data = reflect.Zero(from)
+			} else {
+				data = data.Elem()
+			}
+			return outputHook(from, to, data)
+		case to.Kind() == reflect.Ptr: // foo => *bar
+			// Decoding a non-pointer type to a pointer. Decode as usual and take
+			// a pointer to the result.
+			out, err := outputHook(from, to.Elem(), data)
+			if err != nil {
+				return out, err
+			}
+
+			// If we didn't know what to do with the input, the returned value
+			// will just be the data as-is and it won't have the correct type.
+			if !out.Type().AssignableTo(to.Elem()) {
+				return data, nil
+			}
+
+			result := reflect.New(to.Elem())
+			result.Elem().Set(out)
+			return result, nil
+		}
+
+		return hook(from, to, data)
+	}
+	return
+}
+
+// Builds a reflectHook which unmarshals types using the given unmarshaling
+// scheme. See the unmarshaler type for more information.
+func unmarshalerHook(opts *options) reflectHook {
+	return func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		if !reflect.PtrTo(to).Implements(opts.Unmarshaler.Interface) {
+			return data, nil
+		}
+
+		// The following lines are roughly equivalent to,
+		// 	value := new(foo)
+		// 	err := value.Decode(...)
+		// 	return *value, err
+		value := reflect.New(to)
+		err := opts.Unmarshaler.Unmarshal(value, decodeFrom(opts, data.Interface()))
+		if err != nil {
+			err = fmt.Errorf("could not decode %v from %v: %v", to, from, err)
+		}
+		return value.Elem(), err
+	}
+}
+
+// A reflectHook which decodes strings into time.Durations.
+func durationHook(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
```

Comment:
The implementation is the same as mapstructure's
`StringToTimeDurationHookFunc` except it is compatible with the `reflectHook`
signature. We need to wrap the 'outermost' hook after composition with
`supportPointers` to have pointers inside structs work the way we want, and
for that we need a concrete hook implementation to work with.
`ComposeDecodeHookFunc` composes both, `DecodeHookFuncKind` and
`DecodeHookFuncType`, and does not guarantee what its own signature will be,
so wrapping it to support pointers is going to be difficult.